core/src/peer.rs

```diff
@@ -0,0 +1,208 @@
+use anyhow::{anyhow, bail};
+use multiaddr::{Multiaddr, PeerId};
+use serde::{Deserialize, Serialize};
+use ssi::jws::DecodedJWS;
+
+use crate::{node_id::NodeKey, signer::Signer, DeserializeExt as _, NodeId, SerializeExt as _};
+
+/// Peer entry that is signed and can be shared.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct PeerEntry {
+    id: NodeId,
+    // Number of seconds after UNIX epoch when this entry is no longer valid.
+    expiration: u32,
+    addresses: Vec<Multiaddr>,
+}
+
+impl PeerEntry {
+    /// Construct an entry about a peer with address that is no longer valid after expiration seconds after the
+    /// UNIX epoch.
+    pub fn new(local_id: NodeId, expiration: u32, addresses: Vec<Multiaddr>) -> Self {
+        let peer_id = local_id.peer_id();
+        Self {
+            id: local_id,
+            expiration,
+            addresses: addresses
+                .into_iter()
+                .map(|addr| ensure_multiaddr_has_p2p(addr, peer_id))
```

Comment:
should this be an error if the peer has no addresses?

Reply:
I don't think so. Its not useful to advertise your existence with no addresses, but I don't think its an error.